module mock does not work in some cases in browser mode

### Describe the bug

When mocking the [contentful](https://github.com/contentful/contentful.js) package in browser mode, the mock doesn't seem to apply the first time tests are run and the mocked named export is undefined. Rerunning the tests causes the named export to be a spy, but it does not register being called.

Changing to the `jsdom` environment causes tests to pass.

```
import { createClient as contentfulCreateClient } from "contentful";
import { createClient as localCreateClient } from "../src/contentful.js";
import { vi, expect, describe, it } from "vitest";

const mocks = vi.hoisted(() => ({
  contentfulMock: vi.fn(),
  localMock: vi.fn(),
}));
vi.mock("contentful", () => ({
  createClient: mocks.contentfulMock,
}));
vi.mock("../src/contentful.js", () => ({
  createClient: mocks.localMock,
}));

describe("mocking", () => {
  describe("from contentful", () => {
    it("can make a client", () => {
      // this fails, it's undefined?
      contentfulCreateClient({
        accessToken: "accessToken",
        space: "space",
      });
      expect(mocks.contentfulMock).toHaveBeenCalled();
    });
  });
  describe("from local", () => {
    it("can make a client", () => {
      localCreateClient({
        accessToken: "accessToken",
        space: "space",
      });
      expect(mocks.localMock).toHaveBeenCalled();
    });
  });
});
```

First test run: `TypeError: createClient is not a function`
2nd test run: `AssertionError: expected "spy" to be called at least once`

### Reproduction

https://github.com/kwojcik/vitestsourcemapbug/tree/contentfulMock
```
git checkout contentfulMock
npm install
npm test
```

Rerunning the tests gives a different error

### System Info

```shell
System:
    OS: macOS 13.6.7
    CPU: (12) arm64 Apple M2 Max
    Memory: 6.03 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.2 - ~/.asdf/installs/nodejs/18.18.2/bin/node
    Yarn: 1.22.21 - /opt/homebrew/bin/yarn
    npm: 9.8.1 - ~/.asdf/plugins/nodejs/shims/npm
    pnpm: 9.4.0 - /opt/homebrew/bin/pnpm
    Watchman: 2024.04.01.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 126.0.6478.127
    Safari: 16.6
  npmPackages:
    @vitest/browser: v2.0.0-beta.12 => 2.0.0-beta.12
    @vitest/ui: latest => 2.0.0-beta.12
    vite: latest => 5.3.2
    vitest: v2.0.0-beta.12 => 2.0.0-beta.12
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/browser/src/client/tester/mocker.ts
+++ b/packages/browser/src/client/tester/mocker.ts
@@ -138,16 +138,24 @@ export class VitestBrowserClientMocker {
   public queueMock(id: string, importer: string, factory?: () => any) {
     const promise = rpc()
       .resolveMock(id, importer, !!factory)
-      .then(async ({ mockPath, resolvedId }) => {
+      .then(async ({ mockPath, resolvedId, needsInterop }) => {
         this.ids.add(resolvedId)
         const urlPaths = resolveMockPaths(cleanVersion(resolvedId))
         const resolvedMock
           = typeof mockPath === 'string'
             ? new URL(resolvedMockedPath(cleanVersion(mockPath)), location.href).toString()
             : mockPath
+        const _factory = factory && needsInterop
+          ? async () => {
+            const data = await factory()
+            return { default: data }
+          }
+          : factory
         urlPaths.forEach((url) => {
           this.mocks[url] = resolvedMock
-          this.factories[url] = factory!
+          if (_factory) {
+            this.factories[url] = _factory
+          }
         })
         channel.postMessage({
           type: 'mock',

--- a/packages/browser/src/client/tester/mocker.ts
+++ b/packages/browser/src/client/tester/mocker.ts
@@ -140,10 +140,10 @@ export class VitestBrowserClientMocker {
       .resolveMock(id, importer, !!factory)
       .then(async ({ mockPath, resolvedId }) => {
         this.ids.add(resolvedId)
-        const urlPaths = resolveMockPaths(resolvedId)
+        const urlPaths = resolveMockPaths(cleanVersion(resolvedId))
         const resolvedMock
           = typeof mockPath === 'string'
-            ? new URL(resolvedMockedPath(mockPath), location.href).toString()
+            ? new URL(resolvedMockedPath(cleanVersion(mockPath)), location.href).toString()
             : mockPath
         urlPaths.forEach((url) => {
           this.mocks[url] = resolvedMock
@@ -170,7 +170,7 @@ export class VitestBrowserClientMocker {
           return
         }
         this.ids.delete(resolved.id)
-        const urlPaths = resolveMockPaths(resolved.id)
+        const urlPaths = resolveMockPaths(cleanVersion(resolved.id))
         urlPaths.forEach((url) => {
           delete this.mocks[url]
           delete this.factories[url]
@@ -444,4 +444,9 @@ function resolveMockPaths(path: string) {
   }
 
   return paths
+}
+
+const versionRegexp = /(\?|&)v=\w{8}/
+function cleanVersion(url: string) {
+  return url.replace(versionRegexp, '')
 }

--- a/packages/browser/src/client/tester/msw.ts
+++ b/packages/browser/src/client/tester/msw.ts
@@ -13,7 +13,7 @@ export function createModuleMocker() {
 
   const worker = setupWorker(
     http.get(/.+/, async ({ request }) => {
-      const path = removeTimestamp(request.url.slice(location.origin.length))
+      const path = cleanQuery(request.url.slice(location.origin.length))
       if (!mocks.has(path)) {
         return passthrough()
       }
@@ -112,9 +112,9 @@ function getFactoryExports(id: string) {
 }
 
 const timestampRegexp = /(\?|&)t=\d{13}/
-
-function removeTimestamp(url: string) {
-  return url.replace(timestampRegexp, '')
+const versionRegexp = /(\?|&)v=\w{8}/
+function cleanQuery(url: string) {
+  return url.replace(timestampRegexp, '').replace(versionRegexp, '')
 }
 
 function passthrough() {

--- a/packages/browser/src/node/resolveMock.ts
+++ b/packages/browser/src/node/resolveMock.ts
@@ -1,8 +1,9 @@
-import { existsSync, readdirSync } from 'node:fs'
+import { existsSync, readFileSync, readdirSync } from 'node:fs'
 import { builtinModules } from 'node:module'
 import { basename, dirname, extname, isAbsolute, join, resolve } from 'pathe'
 import type { PartialResolvedId } from 'rollup'
 import type { ResolvedConfig } from 'vitest'
+import type { ResolvedConfig as ViteConfig } from 'vite'
 import type { WorkspaceProject } from 'vitest/node'
 
 export async function resolveMock(
@@ -14,7 +15,9 @@ export async function resolveMock(
   const { id, fsPath, external } = await resolveId(project, rawId, importer)
 
   if (hasFactory) {
-    return { type: 'factory' as const, resolvedId: id }
+    const needsInteropMap = viteDepsInteropMap(project.browser!.vite.config)
+    const needsInterop = needsInteropMap?.get(fsPath) ?? false
+    return { type: 'factory' as const, resolvedId: id, needsInterop }
   }
 
   const mockPath = resolveMockPath(project.config.root, fsPath, external)
@@ -125,4 +128,30 @@ export function isNodeBuiltin(id: string): boolean {
 const postfixRE = /[?#].*$/
 export function cleanUrl(url: string): string {
   return url.replace(postfixRE, '')
+}
+
+const metadata = new WeakMap<ViteConfig, Map<string, boolean>>()
+
+function viteDepsInteropMap(config: ViteConfig) {
+  if (metadata.has(config)) {
+    return metadata.get(config)!
+  }
+  const cacheDirPath = getDepsCacheDir(config)
+  const metadataPath = resolve(cacheDirPath, '_metadata.json')
+  if (!existsSync(metadataPath)) {
+    return null
+  }
+  const { optimized } = JSON.parse(readFileSync(metadataPath, 'utf-8'))
+  const needsInteropMap = new Map()
+  for (const name in optimized) {
+    const dep = optimized[name]
+    const file = resolve(cacheDirPath, dep.file)
+    needsInteropMap.set(file, dep.needsInterop)
+  }
+  metadata.set(config, needsInteropMap)
+  return needsInteropMap
+}
+
+function getDepsCacheDir(config: ViteConfig): string {
+  return resolve(config.cacheDir, 'deps')
 }

--- a/packages/browser/src/node/types.ts
+++ b/packages/browser/src/node/types.ts
@@ -35,6 +35,7 @@ export interface WebSocketBrowserHandlers {
     type: 'factory' | 'redirect' | 'automock'
     mockPath?: string | null
     resolvedId: string
+    needsInterop?: boolean
   }>
   invalidate: (ids: string[]) => void
   getBrowserFileSourceMap: (

--- a/test/browser/fixtures/mocking/mocking-dep.test.ts
+++ b/test/browser/fixtures/mocking/mocking-dep.test.ts
@@ -1,0 +1,12 @@
+import { a } from '@vitest/cjs-lib'
+import { expect, test, vi } from 'vitest'
+
+vi.mock(import('@vitest/cjs-lib'), () => {
+  return {
+    a: 'mocked',
+  }
+})
+
+test('mocking works correctly', () => {
+  expect(a).toBe('mocked')
+})

--- a/test/browser/fixtures/mocking/vitest.config.ts
+++ b/test/browser/fixtures/mocking/vitest.config.ts
@@ -5,6 +5,10 @@ const name =
   process.env.BROWSER || (provider === 'playwright' ? 'chromium' : 'chrome')
 
 export default defineConfig({
+  optimizeDeps: {
+    include: ['@vitest/cjs-lib'],
+    needsInterop: ['@vitest/cjs-lib'],
+  },
   test: {
     browser: {
       enabled: true,

